### PR TITLE
Native Android UI for popup messages and Yes/No dialog

### PR DESCRIPTION
**Pull request by BrettDong**
#### Summary
Interface "Native Android UI for popup messages and Yes/No dialog"

#### Purpose of change
This is a proof of concept that it is feasible to interact with native Android UI components in Java side from Cataclysm's C++ code through JNI. There is a potential to make the game more touch-screen friendly on Android by allowing more menus to interact on these kind of system native touch-screen UI components.

Screenshot_20211201_142539_com cleverraven cataclysmdda experimental
Screenshot_20211201_142602_com cleverraven cataclysmdda experimental
Screenshot_20211201_142627_com cleverraven cataclysmdda experimental

The idea of exploring its feasibility and finally end up having this pull request comes from when I was playing the Android port of _NetHack_. The Android port replaces keyboard inputs with Android system UI components in almost all menus such as player inventory, eating, using items, etc., so the gameplay experience on Android is pretty smooth.

| Role selection | Use an item |
|---|---|
|  |  |

Screenshot_20211201_125429

#### Describe the solution
I added an option `"ANDROID_NATIVE_UI"` that users can choose whether to stick with the original keyboard based UI menus or to use the new Android system menus.

In this pull request I only covered popup messages and yes/no dialogues. I also modified other menus locally but I don't think those changes are stable enough to be production ready so I'm not posting them in this pull request.

The tricky part is JNI and threading. In Cataclysm's C++ side all UI menus are single-threaded and blocking, but Android UI components are designed to be asynchronous and non-blocking. After trying a number of alternatives, the current code is what finally turns out to be working. I use semaphore to block the execution path in Java side to wait for user response, and unblock in Android system menu callback function. This works well. In previous failed attempts, there could be crashes, deadlocks, or UI not redrawing, or touch events not being handled, or user response results not returned to C++ side.

#### Describe alternatives you've considered
An alternative is to make existing UI controls touch-screen aware. For example, let users be able to directly touch on "Yes" or "No" button on the screen:

But the buttons are too small. So maybe need to develop touch-screen variants of all UI controls so touch area of UI elements is bigger, and can directly process raw screen touch events.

So basically this is developing touch screen aware UI controls in Cataclysm's own C++ codebase rather than using Android system UI components.

This approach saves complex boilerplate code to pass data across JNI, and is cross platform not limited to Android.

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
Can you please make this new option to appear in Launcher menu?

**Comment by I-am-Erk** [comment]
Oh my goodness, this is just lovely. I may actually start playing on my phone.

**Comment by ZhilkinSerg** [comment]
APK for testing - 
https://www.dropbox.com/s/hkbiy1q21xzrnip/cataclysmdda-0.f-4191-ge7e79795324-experimental-arm64-v8a-release.apk?dl=0

**Comment by HeYaoDaDa** [comment]
cool
